src/librustc/ty/layout.rs

```diff
@@ -1746,30 +1776,69 @@ impl<'tcx> Deref for TyLayout<'tcx> {
     }
 }
 
-impl<'a, 'gcx, 'tcx> TyLayout<'gcx> {
-    pub fn of(infcx: &InferCtxt<'a, 'gcx, 'tcx>, ty: Ty<'gcx>)
-              -> Result<Self, LayoutError<'gcx>> {
-        let ty = normalize_associated_type(infcx, ty);
+pub trait HasTyCtxt<'tcx>: HasDataLayout {
```

Comment:
If you are defining this trait, maybe move it to `ty::fold` and make `TypeFolder` a subtrait? Can you add a blanket `impl<T: HasTyCtxt> HasDataLayout for T`? Is there any place that needs data layout but does not have a tcx (aka do we need `HasDataLayout`)?